File: Java/src/TreeSort.java
import java.util.ArrayList;

public class TreeSort {

    public class Node{
        Node left;
        Node right;
        int value;
        int count;

        private Node(int val){
            left = null;
            right = null;
            value = val;
            count = 1;
        }
    }


    public Node insert_recursive(Node root, int val){
        /*
         * This function will add the node into the right place, or will increment the counter. (Allow for duplicates)
         */
        if (root == null){
            Node newNode = new Node(val);
            return newNode;
        }
        if (val < root.value){
            root.left = insert_recursive(root.left,val);
        }
        else if (val > root.value){
            root.right = insert_recursive(root.right,val);
        }
        else{
            root.count+=1;
        }
        return root;
    }


    public void sort(Node root, ArrayList<Integer> out){
        /*
         * This is the function that will sort the array using tree sort.
         */
        if (root != null){
            sort(root.left,out);
            for(int i = 0; i<root.count; i++) {
                out.add(root.value);
            }
            sort(root.right,out);
        }
    }

    public static void main(String[] args) {
        Node root = null;
        TreeSort treeSort = new TreeSort();
        ArrayList<Integer> arr = new ArrayList<>();
        int[] ints = new int[]{10,4,3,45,436,65,5,34};
        for (int i:ints){
            root = treeSort.insert_recursive(root,i);
        }
        System.out.println(root.right);
        treeSort.sort(root,arr);
        System.out.println(arr);
    }

}

File: Python/treeSort.py
class Node():
    """
    This is the node class. This holds the information for each node in the tree.
    """
    def __init__(self,val=None):
        self.val = val
        self.left = None
        self.right = None
        self.count = 0


def insert_recursive(root,val):
    """
    This function will insert the value to the tree.
    """
    if root == None:
        root = Node(val)
        root.count+=1
        return root

    if val < root.val:
        root.left = insert_recursive(root.left,val)
    elif val > root.val:
        root.right = insert_recursive(root.right,val)
    else:
        root.count+=1
    return root


def treeSort(root,out):
    """
    This is the tree sort algorithm. Tree sort works by constucting a binary tree and traverses it in order.
    A count has been added to allow for duplicate values.
    Time complexity:
        Best case - O(nlogn)
        Average case - O(nlogn)
        Worst case - O(n^2)
    Space compelxity: O(n)

    Parameters
    ------------
    arr - Array containing numbers or letters.

    Return
    ------------
    arr - Array returned.
    """
    if root != None:
        treeSort(root.left,out)
        for _ in range(root.count):
            out.append(root.val)
        treeSort(root.right,out)
        return out

if __name__ == "__main__":
    root = None
    arr = [21,4,65,2,2,5,5,7]
    out = []
    for i in range(len(arr)):
        root = insert_recursive(root,arr[i])
    treeSort(root,out)
    print(out)
